### PR TITLE
Toggle Vim

### PR DESCRIPTION
**Pull request by rebornix**
#1055 Then you can disable Vim temporarily

cc @jpoon @xconverge @johnfn . I'm happy that we don't have too many dirty checks in code but our package.json is laughing at us.

### CODE REVIEW

File: src/globals.ts [hunk 1]
```diff
@@ -9,4 +9,6 @@ export class Globals {
   public static modeHandlerForTesting: any = undefined;
 
   public static WhitespaceRegExp = new RegExp("^ *$");
+
+  public static active = true;
```
Comment (xconverge) [inline src/globals.ts]: can you add a comment so people know what this means

File: extension.ts [hunk 2]
```diff
@@ -238,15 +242,54 @@ export async function activate(context: vscode.ExtensionContext) {
     registerCommand(context, keybinding.command, () => handleKeyEvent(`${ bracketedKey }`));
   }
 
+  registerCommand(context, 'toggleVim', async () => {
+    Globals.active = !Globals.active;
+    if (Globals.active) {
```
Comment (xconverge) [inline extension.ts]: I imagine somewhere in here we want to do this for each mh we currently have?

Comment (xconverge) [inline extension.ts]: Or does having it just here impact all mh correctly already?

**Comment by xconverge** [comment]
maybe want to clear out the status bar too? or put [vim-disabled] or something

**Comment by jpoon** [comment]
this feels so dirty. so we really need/want this change? the fact that we *can* do it doesn't necessarily mean we *should*. 

i personally would never use this and it seems sufficient to disable this through the extensions UI...

**Comment by johnfn** [comment]
The one use case I know of is snippets, which we don't currently support. I've heard of people restarting their editor and disabling vim just to use a few snippets, which is kind of insane. With this and xconverges work making restarts unnecessary for config changes, I can see this being used by some people. 

Agreed that it's not the idea solution for any individual problem we have though.

**Comment by xconverge** [comment]
It is in VsVim, I have used the vsvim version on numerous occasions.

My use case is a coworker comes over and wants to explain something etc, as soon as they touch the keyboard, crazy stuff happens if they don't use vim. I personally will never use this for myself, it is only for that use case.

**Comment by johnfn** [comment]
@xconverge I think what @jpoon is saying is that this doesn't offer much advantage over just disabling the extension in the sidebar.

**Comment by xconverge** [comment]
Ok so what if I want to interrupt my coworker and type out a function?

I see what you guys are saying, but I don't think it is that dirty, and vscode is slow to disable/reload

**Comment by johnfn** [comment]
He _could_ just press `i` for you, hehe.

I'm ambivalent, but I'm never _really_ against adding new features. 😉  (Oh, except useful ones, of course. _ducks_ )

**Comment by xconverge** [comment]
I have tried pressing i for coworkers, but then inevitably they go to copy paste or something and it just becomes silly for both of us

**Comment by johnfn** [comment]
Haha and then it cuts off the last character and you get in a fighting match that ends in attempting to install spacemacs for 20 minutes?

File: extension.ts (hunk 2 above)
Comment (rebornix) [inline extension.ts]: Nice catch!

File: src/globals.ts (hunk 1 above)
Comment (rebornix) [inline src/globals.ts]: Addressed.

**Comment by jpoon** [comment]
My question is do people do this so often that they need a keyboard shortcut to save them the 2 seconds it would take to do through the UI?

**Comment by rebornix** [comment]
My use case is I'm trying to avoid reload VS Code as possible

* VSCode source code is large and once Code is reloaded, it takes quite a while for TypeScript Language Service to warm up. 
* Tasks will be terminated when reloading Code
* You can't reload as you are in a debugging session. etc, etc.

If I disable Vim for VS Code workspace, sometimes I forgot to bring it back as it's costly.

@jpoon It would be a good feature request to have a new handler called `disableExtension` in `extension.ts` but that requires thorough thinking. I'm not happy with the code either, it's nasty and I can wait for more ppl's feedback or request for this.

**Comment by xconverge** [comment]
I think we should leverage #1425 here and add the vim.use< tag to every binding. Then we can use the Global.isActive flag to set context

**Comment by Chillee** [comment]
I think this would be useful. There are several cases where the plugin as it currently stands has very painful behavior to deal with (namely snippets and folds), and it would be very helpful to be able to disable it temporarily.

**Comment by jpoon** [comment]
Dammit, I don't know how to git :(

I intended to resolve the conflicts, but ended up just creating a new brand-new branch. https://github.com/VSCodeVim/Vim/tree/toggle_vim

**Comment by rebornix** [comment]
@jpoon i'll rebase and resolve the conflicts

**Comment by jpoon** [comment]
I already did the work of resolving the conflicts here: https://github.com/VSCodeVim/Vim/tree/toggle_vim. We can create a new PR with that and close this one? Of you can try cherry-picking commits from there to your current branch.

Sorry, this was more complex than it needed to be.

**Pull request closed without merging**